api/search.go

```diff
@@ -0,0 +1,374 @@
+// Copyright 2018 The WPT Dashboard Project. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package api
+
+import (
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	mapset "github.com/deckarep/golang-set"
+	"github.com/web-platform-tests/wpt.fyi/shared"
+	"google.golang.org/appengine"
+	"google.golang.org/appengine/memcache"
+	"google.golang.org/appengine/urlfetch"
+)
+
+// SearchRunResult is the metadata associated with a particular
+// (test run, test file) pair.
+type SearchRunResult struct {
+	// Passes is the number of test results in a PASS/OK state.
+	Passes int `json:"passes"`
+	// Total is the total number of test results for this run/file pair.
+	Total int `json:"total"`
+}
+
+// SearchResult contains data regarding a particular test file over a collection
+// of runs. The runs are identified externally in a parallel slice (see
+// SearchResponse).
+type SearchResult struct {
+	// Name is the full path of the test file.
+	Name string `json:"name"`
+	// Status is the results data for this file for each relevant run.
+	Status []SearchRunResult `json:"status"`
+}
+
+// SearchResponse contains a response to search API calls, including specific
+// runs whose results were searched and the search results themselves.
+type SearchResponse struct {
+	// Runs is the specific runs for which results were retrieved. Each run, in
+	// order, corresponds to a Status entry in each SearchResult in Results.
+	Runs []shared.TestRun `json:"runs"`
+	// Results is the collection of test results, grouped by test file name.
+	Results []SearchResult `json:"results"`
+}
+
+// summary is the golang type for the JSON format in pass/total summary files.
+type summary map[string][]int
+
+type byName []SearchResult
+
+func (r byName) Len() int           { return len(r) }
+func (r byName) Swap(i, j int)      { r[i], r[j] = r[j], r[i] }
+func (r byName) Less(i, j int) bool { return r[i].Name < r[j].Name }
+
+type readable interface {
+	Get(string) ([]byte, error)
+}
+
+type readWritable interface {
+	readable
+	Put(string, []byte) error
+}
```

Comment:
Is there a reason why we can't just use `io.Reader` & `io.ReadWriter`?

Reply:
I'll can revisit that. My initial attempt to use it seemed more complex because this module then needed to manage `RealData <--> Reader/Writer <--> RealData <--> Cache/Store`. That said, I created these interfaces before I bolted `gzip` support on top, so maybe that code will become simpler with "real" `Reader` `ReadWriter` support.

Reply:
I don't think my first statement was very clear. What I meant to say was some of the APIs we interact with (e.g., memcache) doesn't have a `Reader`/`Writer`-oriented interface, so the wins we should have gotten from "simply decorating" their interface doesn't apply.

Reply:
Not sure about your initial attempt/motivation, but I think what you have right now is quite clear. I like the abstraction of "readable"/"writable" and the interaction between HTTP, gzip & memcache. It's just that these interfaces are almost the same as the standard ones in `io` (if you rename the methods to `Read` & `Write`). And I think it'd be even clearer to use the standard interfaces and this would become a good example of Go's Reader/Writer paradigm.